Mathlib/Algebra/Algebra/Unitize.lean

```diff
@@ -0,0 +1,348 @@
+/-
+Copyright (c) 2023 Jireh Loreaux. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Jireh Loreaux
+-/
+
+import Mathlib.Algebra.Algebra.NonUnitalSubalgebra
+import Mathlib.Algebra.Star.Subalgebra
+import Mathlib.Algebra.Algebra.Unitization
+import Mathlib.Algebra.Star.NonUnitalSubalgebra
+
+/-!
+# Relating unital and non-unital substructures
+
+This file takes unital and non-unital structures and relates them.
+
+## Main declarations
+
+* `NonUnitalSubalgebra.unitization s : Unitization R s →ₐ[R] Algebra.adjoin R (s : Set A)`:
+  where `s` is a non-unital subalgebra of a unital `R`-algebra `A`, this is the natural algebra
+  homomorphism sending `(r, a)` to `r • 1 + a`. This is always surjective.#check
+* `NonUnitalSubalgebra.unitizationAlgEquiv s : Unitization R s ≃ₐ[R] Algebra.adjoin R (s : Set A)`:
+  when `R` is a field and `1 ∉ s`, the above homomorphism is injective is upgraded to
+  an `AlgEquiv`.
+* `Subsemiring.closureEquivAdjoinNat : Subsemiring.closure s ≃ₐ[ℕ] Algebra.adjoin ℕ s`: the
+  identity map between these subsemirings, viewed as `ℕ`-algebras.
+* `Subring.closureEquivAdjoinInt : Subring.closure s ≃ₐ[ℤ] Algebra.adjoin ℤ s`: the
+  identity map between these subsemirings, viewed as `ℤ`-algebras.
+* `NonUnitalSubsemiring.unitization : Unitization ℕ S →ₐ[ℕ] Subsemiring.closure (S : Set R)`:
+  the natural `ℕ`-algebra homomorphism between the unitization of a non-unital subsemiring `S` and
+  its `Subsemiring.closure`. This is just `NonUnitalSubalgebra.unitization S` where we replace the
+  codomain using `NonUnitalSubsemiring.closureEquivAdjoinNat`
+* `NonUnitalSubring.unitization : Unitization ℤ S →ₐ[ℤ] Subring.closure (S : Set R)`:
+  the natural `ℤ`-algebra homomorphism between the unitization of a non-unital subring `S` and
+  its `Subring.closure`. This is just `NonUnitalSubalgebra.unitization S` where we replace the
+  codomain using `NonUnitalSubring.closureEquivAdjoinInt`
+
+-/
+
+section Generic
+
+variable {R S A : Type _} [CommSemiring R] [Semiring A] [Algebra R A] [SetLike S A]
+  [hSA : NonUnitalSubsemiringClass S A] [hSRA : SMulMemClass S R A] (s : S)
+
+/-- The natural `R`-algebra homomorphism from the unitization of a non-unital subalgebra to
+its `Algebra.adjoin`. -/
+def NonUnitalSubalgebra.unitization : Unitization R s →ₐ[R] Algebra.adjoin R (s : Set A) :=
+  Unitization.lift
+    { toFun := fun x : s => (⟨x, Algebra.subset_adjoin x.prop⟩ : Algebra.adjoin R (s : Set A))
+      map_smul' := fun (_ : R) _ => rfl
+      map_zero' := rfl
+      map_add' := fun _ _ => rfl
+      map_mul' := fun _ _ => rfl }
```

Comment:
I wish, but I don't think so. I assume you mean in conjunction with `Unitization.lift`? About all I could do was this:
```lean
def foo :
    s →ₙₐ[R] (Algebra.adjoin R (s : Set A)).toNonUnitalSubalgebra := 
  NonUnitalAlgHom.codRestrict (NonUnitalSubalgebraClass.subtype s) _ fun x ↦ Algebra.subset_adjoin x.property
```
But the problem is you can't `Unitization.lift` this because we need this `Subalgebra.toNonUnitalSubalgebra` bit, and then we no longer know this thing is an `R`-algebra because we "forgot" about 1. So, it's kind of messy.

Reply:
In theory we should be able to use subobject classes, but we lack infrastructure so let's keep it like that for now. I've opened #6335 and #6336 to track what would be needed for that